{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7c825130-fe39-43b3-b997-75a7282b1c82",
   "metadata": {},
   "source": [
    "## Constructing, loading and navigating compacted colored De Bruijn graphs with Pyfrost\n",
    "\n",
    "Pyfrost is a Python library to create and navigate compacted colored De Bruijn graphs with a NetworkX-like API. Behind the scenes, Pyfrost powered by [Bifrost](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02135-8), a memory efficient data structure and method to construct such graphs. This tutorial will show how to create or load existing graphs with Pyfrost, and how to navigate them.\n",
    "\n",
    "### Download test data\n",
    "\n",
    "First, let's download some assemblies as test data. We'll use two *C. portucalensis* genome assemblies as input for our graph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a8e62f64-1ac1-4676-b8cc-71548477c8be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pathlib import Path\n",
    "import ncbi_genome_download as ngd\n",
    "\n",
    "Path(\"_data\").mkdir(exist_ok=True)\n",
    "\n",
    "ngd.download(groups=\"bacteria\", assembly_accessions=\"GCF_001037585.2,GCF_014337275.1\", human_readable=True, file_formats=\"fasta\", output=\"_data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1938ff43-1164-48c6-86f8-70759aa65bc9",
   "metadata": {},
   "source": [
    "### Creating and loading a graph\n",
    "\n",
    "Pyfrost provides the functions, `build`, `build_from_refs`, `build_from_samples` and `load` to create or load a graph. All `build_*` functions accept a `k` and `g` parameter for the k-mer size and minimizer size respectively. The `load` function will obtain the stored k-mer and minimzer size.\n",
    "\n",
    "**WARNING!** The k-mer and minimizer size are set globally. It's not supported to load/create different graphs with different k-mer sizes in the same process!\n",
    "\n",
    "#### Creating a graph from genome assemblies\n",
    "\n",
    "To create a colored De Bruijn graph from a collection of genome assemblies, use the `pyfrost.build_from_refs`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "925dd508-c6ff-44ff-b06d-28bf19020ba6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Citrobacter/sp./BIDMC108/GCF_001037585.2_ASM103758v2_genomic.fna.gz', 'Citrobacter/freundii/MGH279/GCF_014337275.1_ASM1433727v1_genomic.fna.gz']\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pyfrost\n",
    "\n",
    "\n",
    "cwd = Path.cwd()\n",
    "ref_dir = Path(\"_data/human_readable/refseq/bacteria\")\n",
    "os.chdir(ref_dir)\n",
    "\n",
    "# build_from_refs() only supports str objects, no Path objects\n",
    "assembly_list = [str(p) for p in Path().glob(\"**/*.fna.gz\")]\n",
    "print(assembly_list)\n",
    "\n",
    "g = pyfrost.build_from_refs(assembly_list)\n",
    "\n",
    "os.chdir(cwd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a00dd27a-134b-49ac-adcc-f9114812ff74",
   "metadata": {},
   "source": [
    "#### Creating a graph from Illumina reads\n",
    "\n",
    "Pyfrost has a separate function `pyfrost.build_from_samples` to build a graph from read data sets. Since read data often contains sequencing errors, generating a lot of errorneous k-mers occuring only once, this function filters singleton k-mers. If you want to create a graph from both assemblies and reads, use `pyfrost.build`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "017028ec-60f8-43be-adb8-01d48bf20461",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = pyfrost.build_from_samples(['_data/Citr_sp_BIDMC108.subsampled.concat.fq.gz'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "265e5d51-629c-42a4-b776-441d8b9d0153",
   "metadata": {},
   "source": [
    "#### Loading an existing graph created with `Bifrost`\n",
    "\n",
    "If you want to load a graph created using the `Bifrost` CLI tool, use the `pyfrost.load` function, and point it to the GFA file generated by `Bifrost`. Pyfrost only supports **colored** De Bruijn graphs created by Bifrost (so enable the `-c` flag). To demonstrate how this works, let's build an example graph using `Bifrost` first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f261850e-f785-434a-9c1e-795ca26daf24",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Let's build a graph from both the assemblies and the read data using the Bifrost CLI tool\n",
    "ls -1 _data/human_readable/refseq/bacteria/*/*/*/*.fna.gz > _data/ref_list.txt\n",
    "Bifrost build -r _data/ref_list.txt -s _data/Citr_sp_BIDMC108.subsampled.concat.fq.gz -c -i -d -o _data/combined_graph"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b1ad9ca-2936-435f-9e5b-81b3ccd8ff57",
   "metadata": {},
   "source": [
    "In the above example, we first built a text file listing the genome assemblies one per line, gave that file as input to the `Bifrost` command (`-r`), additionally specify the path to a read data set (`-s`), enable colors (`-c`), clip short tips (`-i`), and delete isolated unitigs (`-d`). We specify the output prefix as `_data/combined_graph`, resulting in the files `_data/combined_graph.gfa` and `_data/combined_graph.bfg_colors`. The latter file stores all graph color information.\n",
    "\n",
    "To load the graph in Pyfrost, use the `load` function. You only need to specify the path to the GFA file, the path to bfg_colors file is automatically inferred."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d04f9fff-f359-4125-a844-5fa6247231d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = pyfrost.load('_data/combined_graph.gfa')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4851811-93c6-4e52-a9f3-f7cd7226021f",
   "metadata": {},
   "source": [
    "#### A note on \"colors\"\n",
    "\n",
    "One limitation of Bifrost (and therefore Pyfrost), is that there are a limited number of ways to specify which files or sequences should belong to which colors. In Bifrost, each file is one color. Thus, if you have a paired Illumina data, usually split in two files, we need to concatenate the two FASTQ files to ensure that k-mers from that sample obtain the same colors. On the other hand, in case of genome assemblies, each contig is listed in a single FASTA file, and all contigs will thus get the same color. If you want each contig to have a separate color you'll have to split them in individual files."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2b32ac7-7fd8-4539-8394-545cbbd187c6",
   "metadata": {},
   "source": [
    "### Using the graph\n",
    "\n",
    "The resulting graph object mimics a NetworkX's [DiGraph](https://networkx.org/documentation/stable/reference/classes/digraph.html). Thus many standard algorithms included with NetworkX will function automatically.\n",
    "\n",
    "Due to the double stranded nature of DNA, the graph is *[skew symmetric](https://en.wikipedia.org/wiki/Skew-symmetric_graph)*, i.e., for every node (unitig) $v$, it's reverse complement $\\bar{v}$ is also a node. And for every incoming edge $(u, v)$, an equivalent outgoing edge $(\\bar{v}, \\bar{u})$ exists (and vice versa). Under the hood it's all stored efficiently in memory, but you'll have to take this into account when listing nodes and edges.\n",
    "\n",
    "#### Graph attributes\n",
    "\n",
    "To obtain the k-mer size and color names, use the `graph` attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6b39a806-f9d6-4146-9126-fc815007002c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'color_names': ['_data/Citr_sp_BIDMC108.subsampled.concat.fq.gz',\n",
       "  '_data/human_readable/refseq/bacteria/Citrobacter/freundii/MGH279/GCF_014337275.1_ASM1433727v1_genomic.fna.gz',\n",
       "  '_data/human_readable/refseq/bacteria/Citrobacter/sp./BIDMC108/GCF_001037585.2_ASM103758v2_genomic.fna.gz'],\n",
       " 'k': 31,\n",
       " 'g': 23}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.graph"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e334cb8d-26e7-43a7-9d09-093107664b34",
   "metadata": {},
   "source": [
    "### Access unitigs and k-mers\n",
    "\n",
    "#### Unitigs\n",
    "\n",
    "Each node represents a unitig and is keyed by the first k-mer of that unitig. Thus, if you list the first 10 nodes, you'll get back a list of 10 k-mers representing the unitig's heads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bab8157d-f510-45f8-9a9e-d422fd5d7208",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Kmer 'GCGACCGTTACTCGCTGGGGGTAAGAATTAT'>,\n",
       " <Kmer 'CCGACTTCAAAGCGCAGCTGTGGGGCATCCG'>,\n",
       " <Kmer 'GGATGCCCCACAGCTGCGCTTTGAAGTCGGT'>,\n",
       " <Kmer 'CGTTTTCAGCGTTTGCGTGACGGGTTTAGTA'>,\n",
       " <Kmer 'ACTAAACCCGTCACGCAAACGCTGAAAACGC'>,\n",
       " <Kmer 'GGTTGCACTTGCTGAGCCACGTGGGCGGGCG'>,\n",
       " <Kmer 'GCCCGCCCACGTGGCTCAGCAAGTGCAACCA'>,\n",
       " <Kmer 'GCCTGGACGCGCCGCCGGAGCAGCACGTTGT'>,\n",
       " <Kmer 'CAACGTGCTGCTCCGGCGGCGCGTCCAGGCT'>,\n",
       " <Kmer 'CCATATAGGAATAACGGATTGTAGGCGCCGC'>]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(g.nodes)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e8511d93-26b0-44a9-971e-aa4ab3b535ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "134100"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(g.nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19bf33e0-9d49-40c5-a55f-fff38489265e",
   "metadata": {},
   "source": [
    "`g.nodes` acts as a dict-of-dicts for node attributes and accepts both `Kmer` objects or strings of length `k`. For example, the full unitig sequence of a node can be obtained using the 'unitig_sequence' attribute:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e3b8a7bb-724d-41bc-881c-1dc019cc7700",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCGACCGTTACTCGCTGGGGGTAAGAATTATACGGGCTGGTGGATAAAGCGCAAGGATCGCACGGGATCTTTATTAGATCGTTTAAGCAGTAAAGCATCTTTACTCATTAATTTTTTCAATATGCGCGCTAAAACCTGCAGCACTTCTCCAGGATCGTTTACACTTAGCCGATTCTGGATCATCCTGTGGATAAATCGGGAAGAATCTGTGAGAAACAGAAGATCTCTTTCTCAGTTTACGCTATGATCCGCGGTCCTGATCGTTTCAATGGATCCTGATCGGGTAAAAATTGCAGATAGCAATTCGCACGTCACCCTTTGCATAGGGTCTTGTCGATGTGCGTCAACAATCATGAATGTTTCAGCCTTTGTCATTTATCGACTTTTGTTCGAGTGGAGTCCGCCGTGTCACTTTCGCTTTGGCAGCAGTGTCTTGCCCGATTGCAGGATGAGTTACCAGCCACAGAGTTCAGTATGTGGATACGCCCGTTGCAGGCGGAACTGAGCGATAACACGCTGGCTTTGTATGCGCCAAACCGTTTCGTGCTCGATTGGGTAAGGGATAAATACCTTAATAATATTAATGGATTACTCAATAATTTCTGTGGAGCGGATGCCCCACAGCTGCGCTTTGAAGTCGG'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.nodes['GCGACCGTTACTCGCTGGGGGTAAGAATTAT']['unitig_sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d524a869-35ed-4231-98df-43395ee38512",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "611"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g.nodes['GCGACCGTTACTCGCTGGGGGTAAGAATTAT']['length']  # length in number of k-mers!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "40eca725-b859-4efe-9370-c096f2da7878",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<Kmer 'GCGACCGTTACTCGCTGGGGGTAAGAATTAT'>,\n",
       " <Kmer 'CGGATGCCCCACAGCTGCGCTTTGAAGTCGG'>)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# head and tail k-mers of this unitig\n",
    "g.nodes['GCGACCGTTACTCGCTGGGGGTAAGAATTAT']['head'], g.nodes['GCGACCGTTACTCGCTGGGGGTAAGAATTAT']['tail']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1137fb1-74b2-4c57-8a63-827b09b5cb54",
   "metadata": {},
   "source": [
    "To obtain the reverse complement node $\\bar{v}$ for any node $v$, use the `twin` method, which returns the head k-mer of the reverse complement unitig."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "54be5f55-5b67-4ffb-9ec3-91740e8554c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CCGACTTCAAAGCGCAGCTGTGGGGCATCCG ATAATTCTTACCCCCAGCGAGTAACGGTCGC\n"
     ]
    }
   ],
   "source": [
    "twin = g.nodes['GCGACCGTTACTCGCTGGGGGTAAGAATTAT'].twin()\n",
    "\n",
    "print(g.nodes[twin]['head'], g.nodes[twin]['tail'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a03f600-2b0b-421a-9ec9-480fcf746614",
   "metadata": {},
   "source": [
    "#### Finding specific k-mers\n",
    "\n",
    "To find any specific k-mer (which may or may not be the head k-mer of a unitig), use `g.find`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0937bc36-b373-4336-b8b9-c99fdb5fbb65",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = g.find('AATTATACGGGCTGGTGGATAAAGCGCAAGG')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a1ec79-598a-49a3-a8ca-b49fd2d99806",
   "metadata": {},
   "source": [
    "The resulting mapping object has the same attributes as the unitig attributes obtained using `g.nodes`. There are a few additions, e.g., to obtain the position of the mapped k-mer within the unitig."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8d3307b3-d9a8-43b4-83d6-421237c4a543",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unitig k-mer: GCGACCGTTACTCGCTGGGGGTAAGAATTAT position: 25 length (number of k-mers): 1\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(\"unitig k-mer:\", mapping['head'], \"position:\", mapping['pos'], \"length (number of k-mers):\", mapping['length'])\n",
    "\n",
    "subseq = g.nodes[mapping['head']]['unitig_sequence'][mapping['pos']:mapping['pos']+31]\n",
    "print(subseq == mapping['mapped_sequence'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2200c5f-50bd-4cff-b06b-3451c6dc512d",
   "metadata": {},
   "source": [
    "### Unitig and k-mer colors\n",
    "\n",
    "Each k-mer can have associated \"colors\", which represent the original file sources when the graph was created. To see which samples or genome assemblies contain k-mers of a particular unitig, use the 'colors' attribute. This works for both node attributes as well as k-mer mappings. In case of a node attribute (thus representing a whole unitig), the color set represents the union of all colors of all k-mers of that particular unitig."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d290f981-0ec1-404e-a152-adfd45a07d0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{1, 2}\n",
      "Not present in _data/Citr_sp_BIDMC108.subsampled.concat.fq.gz\n",
      "Present in _data/human_readable/refseq/bacteria/Citrobacter/freundii/MGH279/GCF_014337275.1_ASM1433727v1_genomic.fna.gz\n",
      "Present in _data/human_readable/refseq/bacteria/Citrobacter/sp./BIDMC108/GCF_001037585.2_ASM103758v2_genomic.fna.gz\n"
     ]
    }
   ],
   "source": [
    "from pyfrost import Kmer\n",
    "\n",
    "node = Kmer(\"GGTTGCACTTGCTGAGCCACGTGGGCGGGCG\")\n",
    "\n",
    "print(set(g.nodes[node]['colors']))\n",
    "\n",
    "for c, color_name in enumerate(g.graph['color_names']):\n",
    "    if c in g.nodes[node]['colors']:\n",
    "        print(\"Present in\", color_name)\n",
    "    else:\n",
    "        print(\"Not present in\", color_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "753074b9-fd34-461d-b7dc-693016eb8e94",
   "metadata": {},
   "source": [
    "To obtain the associated colors of the k-mer at a specific position within the unitig, you can use indexing. Each time you use indexing, however, a new proxy object is constructed so it might not be the best for performance-critical applications."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "eab762a4-69e4-4171-8a9a-7050ca0a0264",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{1, 2}\n",
      "{1, 2}\n"
     ]
    }
   ],
   "source": [
    "print(set(g.nodes[node]['colors'][0]))\n",
    "print(set(g.find(node)['colors']))  # equivalent because `node` is the first k-mer of this unitig"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d45e3e1-530d-45ae-9885-e4959402e2eb",
   "metadata": {},
   "source": [
    "### Neighbors and navigation\n",
    "\n",
    "Just as in regular NetworkX DiGraphs, you can use `g.succ` (or `g.adj`) and `g.pred` to access edges to a node's neighbors. `g.successors` (or `g.neighbors`) and `g.predecessors` all work as you'd expect. In addition to these standard NetworkX functions, pyfrost also provides `g.color_restricted_successors` and `g.color_restricted_predecessors`, which will only yield neighbors with a specific color."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7f44f3cb-a1e3-4ab0-bb8f-82b35ce9a18c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Node GCGACCGTTACTCGCTGGGGGTAAGAATTAT ... CGGATGCCCCACAGCTGCGCTTTGAAGTCGG\n",
      "C : GGATGCCCCACAGCTGCGCTTTGAAGTCGGC\n",
      "T : GGATGCCCCACAGCTGCGCTTTGAAGTCGGT\n",
      "\n",
      "Node CCGACTTCAAAGCGCAGCTGTGGGGCATCCG ... ATAATTCTTACCCCCAGCGAGTAACGGTCGC\n",
      "C : TAATTCTTACCCCCAGCGAGTAACGGTCGCC\n",
      "T : TAATTCTTACCCCCAGCGAGTAACGGTCGCT\n",
      "\n",
      "Node GGATGCCCCACAGCTGCGCTTTGAAGTCGGT ... TACTAAACCCGTCACGCAAACGCTGAAAACG\n",
      "C : ACTAAACCCGTCACGCAAACGCTGAAAACGC\n",
      "\n",
      "Node CGTTTTCAGCGTTTGCGTGACGGGTTTAGTA ... ACCGACTTCAAAGCGCAGCTGTGGGGCATCC\n",
      "G : CCGACTTCAAAGCGCAGCTGTGGGGCATCCG\n",
      "\n",
      "Node ACTAAACCCGTCACGCAAACGCTGAAAACGC ... CGCCCGCCCACGTGGCTCAGCAAGTGCAACC\n",
      "A : GCCCGCCCACGTGGCTCAGCAAGTGCAACCA\n",
      "G : GCCCGCCCACGTGGCTCAGCAAGTGCAACCG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "for n, neighbors in g.succ.items():\n",
    "    if i == 5:\n",
    "        break\n",
    "    \n",
    "    print(\"Node\", n, \"...\", g.nodes[n]['tail'])\n",
    "    for v in neighbors:\n",
    "        print(v[-1], \":\", g.nodes[v]['head'])\n",
    "        \n",
    "    print()\n",
    "    i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0c53c6f-5ce5-4d25-94cc-c60cd9f702ed",
   "metadata": {},
   "source": [
    "### Other useful functions\n",
    "\n",
    "#### Paths\n",
    "\n",
    "* `pyfrost.path_sequence(g, path)` - Build to full DNA sequence represented by the given path\n",
    "* `pyfrost.path_nucleotide_length(g, path)` - Calculate the full path in number of nucleotides for the given path\n",
    "* `pyfrost.path_rev_compl(g, path)` - Get the reverse complement path\n",
    "* `pyfrost.path_kmers(g, path)` - Get all k-mers for a given path (also internal unitig k-mers)\n",
    "\n",
    "#### Misc sequence utilities\n",
    "\n",
    "* `pyfrost.kmerize_str(seq)` - K-merize a string\n",
    "* `pyfrost.reverse_complement(seq)` - Get the reverse complement of a sequence"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
